api/src/main/java/edu/cornell/mannlib/vitro/webapp/rdfservice/impl/sparql/RDFServiceSparql.java

```diff
@@ -56,887 +74,855 @@
+            HttpResponse response = context != null ? httpClient.execute(meth, context) : httpClient.execute(meth);
+            try {
+                int statusCode = response.getStatusLine().getStatusCode();
+                if (statusCode > 399) {
+                    log.error("response " + statusCode + " to query. \n");
+                    log.debug("update string: \n" + queryStr);
+                    throw new RDFServiceException("Unable to perform SPARQL UPDATE");
+                }
+
+                try (InputStream in = response.getEntity().getContent()) {
+                    consumer.processResultSet(ResultSetFactory.fromXML(in));
+                }
+            } finally {
+                EntityUtils.consume(response.getEntity());
+            }
+        } catch (IOException | URISyntaxException ioe) {
+            throw new RuntimeException(ioe);
+        }
+    }
+
+    /**
+     * Performs a SPARQL ASK query against the knowledge base. The query may have an embedded graph identifier.
+     *
+     * @param queryStr - the SPARQL query to be executed against the RDF store
+     *
+     * @return boolean - the result of the SPARQL query
+     */
+    @Override
+    public boolean sparqlAskQuery(String queryStr) throws RDFServiceException {
+
+        Query query = createQuery(queryStr);
+        QueryExecution qe = QueryExecutionFactory.sparqlService(readEndpointURI, query);
+
+        try {
+            return qe.execAsk();
+        } finally {
+            qe.close();
+        }
+    }
+
+    protected void rebuildGraphUris() {
+        Thread thread = new VitroBackgroundThread(new Runnable() {
+            public void run() {
+                synchronized (RDFServiceJena.class) {
+                    if (rebuildGraphURICache) {
+                        try {
+                            isRebuildGraphURICacheRunning = true;
+                            Set<String> newURIs = new HashSet<>();
+                            try {
+                                String fastJenaQuery = "SELECT DISTINCT ?g WHERE { GRAPH ?g {} } ORDER BY ?g";
+                                newURIs.addAll(getGraphURIsFromSparqlQuery(fastJenaQuery));
+                            } catch (Exception e) {
+                                log.debug("Unable to use non-standard ARQ query for graph list", e);
+                            }
+                            if (graphURIs.isEmpty()) {
+                                String standardQuery = "SELECT DISTINCT ?g WHERE { GRAPH ?g { ?s ?p ?o } }";
+                                newURIs.addAll(getGraphURIsFromSparqlQuery(standardQuery));
+                            }
+                            updateGraphURIs(newURIs);
+                        } catch (Exception e) {
+                            log.error(e, e);
+                        } finally {
+                            isRebuildGraphURICacheRunning = false;
+                            rebuildGraphURICache = false;
```

Comment:
I feel like something is missing here.  What happens if a new graph is added while the rebuild thread is running?  It looks like graphs could be missed since the dirty flag is cleared at the end of the rebuild.  Wouldn't it be better to clear it at the beginning so that it can be re-triggered while the thread is running?

Reply:
Thank you. I added a commit to fix that.